### DISCUSSION TITLE
Read file from ESP32

### DISCUSSION THREAD

**Discussion by ioagr** [nanoframework/Home#1530] [Q&A] [answered]
hi everyone,

I need help with reading files from ESP32 on Windows. I use nf-debugger to write a software settings file to the ESP32 flash and then read that file when the ESP32 boots. To write the file I use the "DeleteStorageFile" and "AddStorageFile" functions from DebugEngine, but there is no function in DebugEngine to read the file if I want to change the settings. I am currently using Debug.WriteLine to send the content of the file from the ESP32 to the Windows application every 5 seconds, but it would be better to read the file only on demand. 
So is there a way to read the file from ESP32 using nf-debugger other than using "Debug.WriteLine"?

```
   SelectedDevice.DebugEngine.DeleteStorageFile("I:\\sSett");
   SelectedDevice.DebugEngine.AddStorageFile("I:\\sSett", Encoding.UTF8.GetBytes(sett));
   SelectedDevice.DebugEngine.RebootDevice(RebootOptions.NormalReboot);
```

One more thing I want to mention: If I compile the code in Release, the "Debug.WriteLine" function stops working, so I have to compile it in Debug to be able to use this function.

And sorry for my bad english.

**Edit:** Using System.IO.SerialPort on ESP32 is not an option for me because all ESP32 pins are used. To open the SerialPort from nanoFramework on default RX/TX pins is also not an option because I can not upload the nanoframework projects anymore if the port is used for something else...

**Edit 2:** I managed to do it like this:

nanoFramework code:

```
public static void Main()
{
       new Thread(VSmessageThread).Start();
}

 static void VSmessageThread()
 {
     while (true)
     {
         Thread.CurrentThread.Suspend();
         Console.WriteLine("Debug: Thread Resumed");
         // Send settings
     }
 }

```

Windows C# code:

 ```
private void BtnConnect_Click(object sender, EventArgs e)
 { 
         SelectedDevice = SelectDeviceMethod();
         if (SelectedDevice.DebugEngine == null)
         {
             SelectedDevice.CreateDebugEngine(NanoSerialDevice.SafeDefaultTimeout);
         }
         SelectedDevice.DebugEngine.OnMessage += DebugEngine_OnMessage;
         SerialConnected = selectedDevice.DebugEngine.Connect(5000, true, false); 
 }

private void DebugEngine_OnMessage(nanoFramework.Tools.Debugger.WireProtocol.IncomingMessage message, string text)
{
      //Retrive settings from "text" parameter.
}
void GetSettings()
{
     var mtRes = SelectedDevice.DebugEngine.GetThreads();
     foreach(var th in mtRes)
     {
         foreach(var mtName in th.m_calls)
         {
              if (mtName.Contains("VSmessageThread"))
              {
                  SelectedDevice.DebugEngine.ResumeThread(th.m_pid); 
              }
         }
     } 
}
```

**Comment by Ellerbach** [comment, marked as answer]
To read the file, you can easily write a small program that will download and read the file. For security reasons we've decided not to implement a too easy way to read the files at flash time.

> Release, the "Debug.WriteLine" function stops working

Yes, because that's for debug :-) Use `Consile.WriteLine`instead ;-) Those always stays regardeless if you are on Debug or Release modes.

**Reply by ioagr** [reply]
> Use Consile.WriteLine instead ;-)

I didn't even think that the "Console.WriteLine" function can be used without a display. This is nice. Also "Console.ReadLine" will be nice to be implemented. This way we will be able to send messages back and forth between the ESP32 and a Windows application using the default serial port.

> For security reasons we've decided not to implement a too easy way to read the files at flash time.

Maybe I have to look more into the nf-debugger library to see how it's implemented...

Thanks for the answer.
